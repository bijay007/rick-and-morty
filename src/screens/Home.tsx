import React, { useState, useEffect } from 'react';
import {
  View, Text, Image, FlatList, StyleSheet,
} from 'react-native';
import { Card, CardItem, Left, Right, Separator } from 'native-base';

interface CharacterInfo {
  name: string,
  status: string,
  image: string
}

function Home() {
  const [isLoading, fetching] = useState(true);
  const [characterList, getList] = useState([]);
  useEffect(() => {
    getCharacters()
  }, [characterList]);

  const getCharacters = async () => {
    try {
      const response = await fetch('https://rickandmortyapi.com/api/character/');
      const json = await response.json();
      const data: [CharacterInfo] = json.results.map(({ name, status, image }) => ({ name, status, image }));
      fetching(false);
      getList(data);
    } catch (err) {
      console.error(err);
    }
  };

  return !isLoading && (
     <View style={styles.wrapper}>
      <FlatList
        data={characterList}
        renderItem={({ item }) => (
          <Card transparent>
            <CardItem>
              <Left style={{flex: 2}}>
                <Image
                  style={styles.image}
                  source={{uri: item.image}}
                  resizeMode={'cover'}
                />
              </Left>
              <Right style={styles.right}>
                <Text>{item.name}</Text>
                <Text>Status: {item.status}</Text>
              </Right>
            </CardItem>
          </Card>
        )}
        keyExtractor={(item, index) => `${item.name}_${index}`}
        removeClippedSubviews={true}
        initialNumToRender={5}
        maxToRenderPerBatch={5}
        windowSize={19}
        ItemSeparatorComponent={() => <Separator style={{height: 4}} />}
      />
    </View>
  );
}

export default Home;

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    width: '100%',
  },
  right: {
   flex: 1,
   alignItems: 'flex-start',
   justifyContent: 'space-around',
  },
  image: {
    height: 150,
    width: 180,
  },
});
