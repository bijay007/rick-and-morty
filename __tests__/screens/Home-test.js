import React from 'react';
import * as renderer from 'react-test-renderer';
import Home from '../../src/screens/Home';

describe('Home', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });
  it('renders correctly', async () => {
    const tree = await renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('contains two inner <Row> children elements', async () => {
    const tree = await renderer.create(<Home />).toJSON();
    expect(tree.children[0].children.length).toBe(1);
  });
});
