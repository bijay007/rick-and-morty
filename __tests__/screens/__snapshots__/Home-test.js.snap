// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "paddingTop": 47,
    }
  }
>
  <View
    style={
      Object {
        "height": 266.8,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "column",
          "height": 150,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Object {
              "flex": 1,
              "height": "90%",
              "paddingVertical": 10,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "flex": 1,
              "height": 1334,
            }
          }
        >
          <View
            onLayout={[Function]}
          >
            <View
              rounded={true}
              searchBar={true}
              style={
                Object {
                  "backgroundColor": "#F8F8F8",
                  "borderBottomColor": "#a7a6ab",
                  "borderBottomWidth": 0.5,
                  "elevation": 3,
                  "flexDirection": "row",
                  "height": 64,
                  "justifyContent": "center",
                  "left": 0,
                  "paddingLeft": 6,
                  "paddingRight": 10,
                  "paddingTop": 18,
                  "right": 0,
                  "shadowColor": undefined,
                  "shadowOffset": undefined,
                  "shadowOpacity": undefined,
                  "shadowRadius": undefined,
                  "top": 0,
                }
              }
            >
              <View
                accessible={true}
                isTVSelectable={true}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "backgroundColor": "#CECDD2",
                    "borderBottomWidth": 1,
                    "borderColor": "transparent",
                    "borderLeftWidth": 0,
                    "borderRadius": 25,
                    "borderRightWidth": 0,
                    "borderTopWidth": 0,
                    "borderWidth": undefined,
                    "flex": 1,
                    "flexDirection": "row",
                    "height": 30,
                    "justifyContent": "flex-start",
                    "marginLeft": 2,
                    "opacity": 1,
                  }
                }
              >
                <Text />
                <TextInput
                  allowFontScaling={true}
                  editable={true}
                  onChangeText={[Function]}
                  placeholder="Search"
                  placeholderTextColor="#575757"
                  rejectResponderTermination={true}
                  style={
                    Object {
                      "alignSelf": "center",
                      "color": "#000",
                      "flex": 1,
                      "fontSize": 17,
                      "height": 30,
                      "lineHeight": null,
                      "paddingLeft": 5,
                      "paddingRight": 5,
                      "top": 1.5,
                    }
                  }
                  underlineColorAndroid="rgba(0,0,0,0)"
                />
                <Text />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="small"
      />
    </View>
  </View>
</View>
`;
